File: services/audit-compliance-service/src/handlers/gdpr.rs
use auth_middleware::layer::AuthUser;
use axum::{Json, extract::State};

use crate::{
    AppState,
    domain::{gdpr, security},
    handlers::{ServiceResult, bad_request, db_error, load_events},
    models::compliance_report::{GdprExportPayload, GdprExportRequest},
};

pub async fn export_subject_data(
    State(state): State<AppState>,
    AuthUser(claims): AuthUser,
    Json(request): Json<GdprExportRequest>,
) -> ServiceResult<GdprExportPayload> {
    if !security::can_access_subject(&claims, &request.subject_id) {
        return Err(bad_request(
            "session scope does not allow this subject export",
        ));
    }

    let events = security::filter_events_for_claims(
        load_events(&state.db)
            .await
            .map_err(|cause| db_error(&cause))?,
        &claims,
    );
    Ok(Json(gdpr::export_payload(&request, &events)))
}

File: services/audit-compliance-service/internal/handlers/gdpr.go
// GDPR subject-export handler — mirrors `handlers/gdpr.rs`.

package handlers

import (
	"encoding/json"
	"net/http"

	authmw "github.com/openfoundry/openfoundry-go/libs/auth-middleware"
	"github.com/openfoundry/openfoundry-go/services/audit-compliance-service/internal/domain/gdpr"
	"github.com/openfoundry/openfoundry-go/services/audit-compliance-service/internal/domain/security"
	"github.com/openfoundry/openfoundry-go/services/audit-compliance-service/internal/models"
)

// ExportSubjectData ports `handlers::gdpr::export_subject_data`.
func (h *Handlers) ExportSubjectData(w http.ResponseWriter, r *http.Request) {
	claims, ok := authmw.FromContext(r.Context())
	if !ok {
		writeJSONErr(w, http.StatusUnauthorized, "authentication required")
		return
	}
	var body models.GdprExportRequest
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeJSONErr(w, http.StatusBadRequest, err.Error())
		return
	}
	if body.SubjectID == "" {
		writeJSONErr(w, http.StatusBadRequest, "subject_id is required")
		return
	}
	if !security.CanAccessSubject(claims, body.SubjectID) {
		writeJSONErr(w, http.StatusBadRequest, "session scope does not allow this subject export")
		return
	}
	all, err := h.Repo.ListAuditEvents(r.Context(), 1000)
	if err != nil {
		writeJSONErr(w, http.StatusInternalServerError, "database operation failed")
		return
	}
	events := security.FilterEventsForClaims(all, claims)
	writeJSON(w, http.StatusOK, gdpr.ExportPayload(&body, events))
}
